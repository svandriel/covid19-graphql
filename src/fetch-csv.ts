import chalk from 'chalk';
import csvParse from 'csv-parse';
import fetch, { FetchError } from 'node-fetch';

export interface CsvContent extends Array<Record<string, string>> {}

export async function fetchCsv(url: string): Promise<CsvContent> {
    return new Promise<CsvContent>(async (resolve, reject) => {
        try {
            console.log(`Fetching ${chalk.cyan(url)}`);
            const response = await fetch(url);
            if (response.status !== 200) {
                throw new FetchError(`Got HTTP ${response.status} on ${url} (expected HTTP 200)`, 'ENOTFOUND');
            }
            const text = await response.text();
            csvParse(
                text,
                {
                    columns: true,
                    delimiter: ',',
                    quote: '"',
                    bom: true
                },
                (err, output) => {
                    if (err) {
                        reject(err);
                    } else {
                        resolve(output);
                    }
                }
            );
        } catch (err) {
            reject(err);
        }
    });
}
