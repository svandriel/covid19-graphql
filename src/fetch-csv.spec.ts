import nock from 'nock';

import { fetchCsv } from './fetch-csv';

describe('fetchCsv', () => {
    beforeEach(() => {
        nock('https://example.com')
            .get('/file.csv')
            .reply(200, `One,Two\n1,2\n3,4`);
        nock('https://example.com')
            .get('/timeout.csv')
            .replyWithError('Timeout');
        nock('https://example.com')
            .get('/server-error.csv')
            .reply(500, 'Server error');
        nock('https://example.com')
            .get('/malformed.csv')
            .reply(200, `One,Two\n1,"2\n3,4`);
    });

    it('works', async () => {
        try {
            const result = await fetchCsv('https://example.com/file.csv');
            expect(result).toMatchSnapshot();
        } catch (err) {
            fail(err);
        }
    });

    it('fails on invalid CSV', async () => {
        try {
            await fetchCsv('https://example.com/malformed.csv');
            fail('Expected error');
        } catch (err) {
            expect(err.message).toBe('Quote Not Closed: the parsing is finished with an opening quote at line 3');
        }
    });

    it('fails on timeout', async () => {
        try {
            await fetchCsv('https://example.com/timeout.csv');
            fail('Expected error');
        } catch (err) {
            expect(err.message).toBe('request to https://example.com/timeout.csv failed, reason: Timeout');
        }
    });

    it('fails on server error', async () => {
        try {
            await fetchCsv('https://example.com/server-error.csv');
            fail('Expected error');
        } catch (err) {
            expect(err.message).toBe('Got HTTP 500 on https://example.com/server-error.csv (expected HTTP 200)');
        }
    });
});
