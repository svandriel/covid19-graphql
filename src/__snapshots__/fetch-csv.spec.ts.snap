// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fetchCsv works 1`] = `
Array [
  Object {
    "One": "1",
    "Two": "2",
  },
  Object {
    "One": "3",
    "Two": "4",
  },
]
`;
